config.py

```diff
@@ -742,157 +754,184 @@ def check_yaml(conf_path):
+    instances = check_config['instances']
+    try:
         try:
-            check_module = imp.load_source('checksd_%s' % check_name, check)
-        except Exception, e:
-            traceback_message = traceback.format_exc()
-            # There is a configuration file for that check but the module can't be imported
-            init_failed_checks[check_name] = {'error':e, 'traceback':traceback_message}
-            log.exception('Unable to import check module %s.py from checks.d' % check_name)
-            continue
+            check = check_class(check_name, init_config=init_config,
+                            agentConfig=agentConfig, instances=instances)
+        except TypeError, e:
+            # Backwards compatibility for checks which don't support the
+            # instances argument in the constructor.
+            check = check_class(check_name, init_config=init_config,
+                            agentConfig=agentConfig)
+            check.instances = instances
+    except Exception, e:
+        log.exception('Unable to initialize check %s' % check_name)
+        traceback_message = traceback.format_exc()
+        return {}, {check_name: {'error':e, 'traceback':traceback_message}}
+    else:
+        return {check_name: check}, {}
 
-        # We make sure that there is an AgentCheck class defined
-        check_class = None
-        classes = inspect.getmembers(check_module, inspect.isclass)
-        for _, clsmember in classes:
-            if clsmember == AgentCheck:
-                continue
-            if issubclass(clsmember, AgentCheck):
-                check_class = clsmember
-                if AgentCheck in clsmember.__bases__:
-                    continue
-                else:
-                    break
+def _update_python_path(check_config):
+    # Add custom pythonpath(s) if available
+    if 'pythonpath' in check_config:
+        pythonpath = check_config['pythonpath']
+        if not isinstance(pythonpath, list):
+            pythonpath = [pythonpath]
+        sys.path.extend(pythonpath)
 
-        if not check_class:
-            log.error('No check class (inheriting from AgentCheck) found in %s.py' % check_name)
-            continue
+def load_check_directory(agentConfig, hostname):
+    ''' Return the initialized checks from checks.d, and a mapping of checks that failed to
+    initialize. Only checks that have a configuration
+    file in conf.d will be returned. '''
+    from checks import AGENT_METRICS_CHECK_NAME
+
+    initialized_checks = {}
+    init_failed_checks = {}
+    deprecated_checks = {}
+    agentConfig['checksd_hostname'] = hostname
+    osname = get_os()
+
+    deprecated_checks.update(_deprecated_configs(agentConfig))
 
-        # Look for the per-check config, which *must* exist
-        if not check_config.get('instances'):
-            log.error("Config %s is missing 'instances'" % conf_path)
+    all_configs_paths = _all_configs_paths(osname, agentConfig)
```

Comment:
Not sure it makes sense for `all_config_paths` to be an list of lists instead of simply a list.